Add changes for windows arm64 and update dependencies

HellAholic opened the pull request:
This pull request introduces several updates to enable Windows ARM build compatibility, dependency management, and package versioning. The main changes include adding support for Windows ARM builds, refining how platform-specific dependencies are handled (especially for `pynavlib`), updating the `cpython` version, and cleaning up dependency definitions for `shapely`.

**Windows build and dependency management:**

* Added support for self-hosted Windows ARM runners in the workflow by updating `.github/workflows/windows.yml`.
* Ensured `pynavlib` is only included for Windows x64 builds by updating `conandata.yml` and modifying the PyInstaller spec generation in `conanfile.py` to exclude `pynavlib` on Windows ARM64. [[1]](diffhunk://#diff-545553e7b8144a900437cd74842021b8640d473390f8a836db86ad55d85dd069R555-R557) [[2]](diffhunk://#diff-63f09721ade419d29886e8190ffa3c9c756baa9bef8802a7f2aeabf3b545164fR552-R555)

**Dependency updates:**

* Upgraded the `cpython` dependency from version 3.12.2 to 3.12.7 in `conanfile.py`.
* Added `shapely/2.0.6` to the conan requirements in `conandata.yml` and removed its pip hashes, consolidating its management under Conan. [[1]](diffhunk://#diff-545553e7b8144a900437cd74842021b8640d473390f8a836db86ad55d85dd069R12) [[2]](diffhunk://#diff-545553e7b8144a900437cd74842021b8640d473390f8a836db86ad55d85dd069L413-L457)

**Add test Workflow management to be able to validate the changes on branch**

* Updated the workflow to use a specific branch/tag (`CURA-12814`) for the Windows installer job for testing windows ARM.

HellAholic committed:
Add arm64 C++20 profile and update dependencies

Introduced a new arm64_cpp20_profile for Windows builds with C++20 and updated several dependency configurations. Added 'shapely/2.0.6' to requirements, removed 'pynavlib' from requirements and pip lists, and updated cpython to version 3.12.7 in conanfile.py.

wawanbreton commented on arm64_cpp20_profile:
This should be in a profile in `conan-config`

wawanbreton commented on conandata.yml:
```diff
@@ -127,7 +128,6 @@ pyinstaller:
         - "PyQt6.sip"
         - "stl"
         - "keyrings.alt"
-        - "pynavlib"
```
Are we really ok with doing this ?

HellAholic committed:
Remove the local generated profile

Jinja template is updated so this file should now be automatically generated

HellAholic committed:
modify pynavlib package for Windows x64 and add back the hashes

Added pynavlib version 0.9.4 with corresponding hashes to pip_requirements_core for Windows x64 and MacOS platforms. This enables platform-specific installation of pynavlib where supported.

HellAholic committed:
Add pynavlib to PyInstaller hidden imports

Included 'pynavlib' in the PyInstaller hiddenimports list. On Windows ARM64, 'pynavlib' is excluded from collect_all due to lack of ARM64 wheels.

wawanbreton committed:
Merge branch 'main' into CURA-12814

HellAholic committed:
Update windows.yml

HellAholic committed:
Merge branch 'main' into CURA-12814

HellAholic committed:
Update workflow references and ARM option

Changed workflow references from 'main' to 'CURA-12814' branch for cura-installers.yml in find-packages.yml and installers.yml. Updated the Windows workflow to use 'self-hosted-Windows-ARM64' instead of 'self-hosted-Windows-ARM'.

HellAholic committed:
Merge branch 'main' into CURA-12814

HellAholic committed:
Merge branch 'main' into CURA-12814

HellAholic committed:
Bump pysavitar and shapely versions

Update conandata.yml to use newer dependency versions: pysavitar -> 5.12.0-alpha.0 and shapely -> 2.1.2. These bumps align the package dependencies with the latest releases referenced by the recipe.

HellAholic committed:
Update conandata.yml

HellAholic committed:
Add scipy to pyinstaller requirements

Include "scipy" in the pyinstaller requires list in conandata.yml so SciPy is collected/included when building the package with PyInstaller. This ensures SciPy-dependent modules are packaged correctly.

HellAholic committed:
Merge branch 'main' into CURA-12814

HellAholic committed:
Fix add printer list view issues updating to PyQt 6.10.0

Replace in-place mutations and self-assign hacks for base.currentSections with creation and assignment of new Set instances to ensure QML reactivity.
Changes in AddLocalPrinterScrollView.qml: initialize currentSections with a new Set on Component.onCompleted; when applying the search filter build and assign a new Set of sections instead of clearing/adding and self-assigning; and update the section toggle handler to operate on a copied Set and assign it back.
Removes redundant self-assignment lines and streamlines section collection logic.

HellAholic committed:
Ensure build volume initialized before arrange

Fix a race where BuildVolume dimensions could remain zero because globalContainerStackChanged/activeQualityChanged fire before BuildVolume is constructed. Call _onStackChanged() during initialization and re-trigger it from _onEngineCreated() (with a warning) if the volume AABB is still None so dimensions are populated and rebuild can succeed. In CuraApplication._arrangeAll, handle a None build-volume bounding box by logging a warning, requesting a rebuild, and aborting arrange to avoid comparing against an uninitialized volume.

HellAholic committed:
Simplify comments and clarify rebuild warnings

Clean up and shorten explanatory comments in BuildVolume related to queuing the stack-change timer, and simplify the warning text when rebuild() returns early or the volume bounding box is missing. These edits reduce verbose explanations about signal ordering/race conditions while keeping behavior unchanged.

HellAholic committed:
Merge branch 'main' into CURA-12814

HellAholic committed:
Merge branch 'main' into CURA-12814

HellAholic committed:
Fix perpetual warning state when loading a 3mf

When a 3mf loaded and the machine needs to re-initialize, the material change and the variant change do not trigger the re-evaluation as we previously only tracked the extruderStack with loose signal tracking. With the new qt version, that flow no longer works as a new binding model is in place, so we need explicit bindings to trigger the re-evaluation for material / variant change.

HellAholic committed:
Fix content overflow in the signin pop up

The Item in AccountDetails.qml had no implicitWidth/implicitHeight, so it defaulted to 0. The Popup in AccountWidget.qml sizes itself from contentItem.implicitWidth/Height — with both being 0, the background rectangle got zero size while the loaded GeneralOperations content still rendered visually, causing the text to appear outside the background.

HellAholic committed:
Fix content overflow open file in prepare stage

The Item in PrepareMenu.qml had no implicitWidth/implicitHeight, so it defaulted to 0. Now the height properly calculates when the user is signed in and there is a second option to open from Digital library present

HellAholic committed:
Remove the explicit height override

Size is now handled with the change in: 81960ad1b3ccdceb240fa3470ec85e296f29bd8b

The explicit height override was causing the hover state to cover the lower border.

HellAholic committed:
Merge branch 'main' into CURA-12814

HellAholic committed:
Merge branch 'main' into CURA-12814

HellAholic committed:
Ensure we don't get a gap at the top when re-rendring the scene

When a new printer is added, the view re-renders, the menu being rendered natively but its wrapper will still get the height and apply a cause a gap.

HellAholic committed:
Filter the addition of objects based on shouldBeVisible

On the native menu for osx, there are unintended plugins showing in the extension list, try to Handle the unintended extensions showing up

HellAholic committed:
Merge branch 'main' into CURA-12814

HellAholic commented on .github/workflows/find-packages.yml:
```diff
@@ -49,7 +49,7 @@ jobs:
     needs: find-packages
     if: ${{ inputs.start_builds == true && needs.find-packages.outputs.discovered_packages != '' }}
-    uses: ultimaker/cura-workflows/.github/workflows/cura-installers.yml@main
+    uses: ultimaker/cura-workflows/.github/workflows/cura-installers.yml@CURA-12814
```
```suggestion
    uses: ultimaker/cura-workflows/.github/workflows/cura-installers.yml@main
```
Apply suggestion after the cura-workflows branch is merged to main

HellAholic commented on .github/workflows/installers.yml:
```diff
@@ -29,7 +29,7 @@ on:
   installers:
     name: Create installers
-    uses: ultimaker/cura-workflows/.github/workflows/cura-installers.yml@main
+    uses: ultimaker/cura-workflows/.github/workflows/cura-installers.yml@CURA-12814
```
```suggestion
    uses: ultimaker/cura-workflows/.github/workflows/cura-installers.yml@main
```
Apply suggestion after the cura-workflows branch is merged to main

HellAholic commented on .github/workflows/windows.yml:
```diff
@@ -34,11 +34,12 @@ on:
 jobs:
   windows-installer:
-    uses: ultimaker/cura-workflows/.github/workflows/cura-installer-windows.yml@main
+    uses: ultimaker/cura-workflows/.github/workflows/cura-installer-windows.yml@CURA-12814
```
```suggestion
    uses: ultimaker/cura-workflows/.github/workflows/cura-installer-windows.yml@main
```
Apply suggestion after the cura-workflows branch is merged to main

HellAholic committed:
Separate the arm/x64

HellAholic commented on .github/workflows/windows-arm.yml:
```diff
@@ -0,0 +1,48 @@
+
+jobs:
+  windows-installer:
+    uses: ultimaker/cura-workflows/.github/workflows/cura-installer-windows-arm.yml@CURA-12814
```
```suggestion
    uses: ultimaker/cura-workflows/.github/workflows/cura-installer-windows-arm.yml@main
```

HellAholic committed:
Add all unused plugins/dlls to the blacklist

We are not using any of these files in the application, so removing them reduces package size.

HellAholic committed:
change target branch for workflows

HellAholic committed:
remove the dbus from blacklist

It's in use in linxu, can be manually removed for windows/macos during packaging

casperlamboo commented on resources/qml/Menus/ExtensionMenu.qml:
```diff
@@ -39,7 +39,7 @@ Cura.Menu
             id: sub_menu
             title: model.name
-            shouldBeVisible: actions !== undefined
+            shouldBeVisible: actions != null
```
```suggestion
            shouldBeVisible: !!actions
```

HellAholic committed:
Switch the approach for filtering and use the jinja template to remove the unused packages

HellAholic committed:
Update UltiMaker-Cura.spec.jinja

HellAholic committed:
Revert jinja changes for filtering

Reverts the changes in:
3175f7c626afa141ec0a251697bda5149a8a73ce
&
438ab1632ff668f14848720c7b9ee804986637fe

HellAholic committed:
Update the blacklist

Extend the blacklist pattern
Change the conanfile.py to accommodate the new patterns introduced

HellAholic committed:
Update conandata.yml

rburema commented on conandata.yml:
```diff
@@ -591,6 +548,9 @@ pip_requirements_core:
         - sha256:d162dc04946d704503b2edc4d55f3dba5c1d539ead017afa00142c38b9885755
+
+  Windows_x64:
+    # pynavlib is only available for Windows x64, not ARM64
```
We need to make sure customers know they can't use the 3D mouse on Win ARM64 builds then.

HellAholic replied:
I think we can ask the 3dconnexion to provide the package for arm, but it was a bit premature to ask them if we don't ship a build.

rburema commented on resources/qml/Menus/ConfigurationMenu/ConfigurationMenu.qml:
```diff
@@ -61,7 +61,12 @@ Cura.ExpandablePopup
-                    property bool valueWarning: !Cura.ExtruderManager.getExtruderHasQualityForMaterial(extruderStack)
+                    property bool valueWarning: {
+                        // Explicitly read material and variant to register them as QML binding dependencies.
+                        var _material = extruderStack ? extruderStack.material : null
+                        var _variant = extruderStack ? extruderStack.variant : null
```
Is that also true if they're in a scope? It could be, but I thought that `{...}` would destroy bindings.

HellAholic replied:
So due to the Qt changes, we need to explicitly register the bindings, otherwise it is not triggering an update on the change.
The `{ ... }` would be breaking when using with a static value I believe.

rburema commented on resources/qml/Menus/ExtensionMenu.qml:
```diff
@@ -58,7 +58,7 @@ Cura.Menu
-        onObjectAdded: function(index, object) { extensionMenu.insertMenu(index, object) }
-        onObjectRemoved: function(index, object) { extensionMenu.removeMenu(object)}
+        onObjectAdded: function(index, object) { if (object.shouldBeVisible) { extensionMenu.insertMenu(index, object) } }
+        onObjectRemoved: function(index, object) { extensionMenu.removeMenu(object) }
```
I assume this will also work if the object wasn't visible when added? In that hopefully, the menu ignores irrelevant 'remove' calls.

HellAholic replied:
Yes, tried with the visibility, but this section is for the macos native menu, and potential follow up work might change this in: CURA-13240 (as either revert/exclusion, or implementation changes to work as expected with the visible/should be visible properties, similar issue in the File menu)

rburema reviewed (approved):
LGTM

HellAholic committed:
code-review

casperlamboo committed:
Simplify blacklisting logic

NP-13236

HellAholic committed:
Merge pull request #21677 from Ultimaker/CURA-13236

Add all unused plugins/dlls to the blacklist

HellAholic committed:
Apply suggestions from code review

Co-authored-by: HellAholic <alireza.doustdar@gmail.com>
Co-authored-by: Casper Lamboo <casperlamboo@gmail.com>
